theseus/embodied/robot/link.py

```diff
@@ -0,0 +1,75 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import abc
+from typing import List, Optional, Any
+import torch
+
+
+class Link(abc.ABC):
+    def __init__(
+        self,
+        name: str,
+        id: int = -1,
+        parent: Any = None,
+        children: Optional[List[Any]] = None,
+        dtype: Optional[torch.dtype] = None,
+    ):
+        self._name = name
+        self._id = id
+        self._parent = parent
+        self._children = children if children else []
+        self._ancestors: List[Link] = []
+        self._dtype = dtype
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def id(self) -> int:
+        return self._id
+
+    @property
+    def parent(self) -> Any:
+        return self._parent
+
+    @property
+    def children(self) -> List[Any]:
+        return self._children
+
+    @property
+    def ancestors(self) -> List["Link"]:
+        return self._ancestors
+
+    @property
+    def dtype(self) -> torch.dtype:
+        return self._dtype
+
+    def set_id(self, id: int):
+        self._id = id
+
+    def set_parent(self, parent: Any):
+        self._parent = parent
+
+    def set_children(self, children: List[Any]):
+        self._children = children
+
+    def set_ancestors(self, ancesotrs: List["Link"]):
+        self._ancestors = ancesotrs
+
+    def update_ancestors(self):
+        joint = self.parent
+        self._ancestors = []
+        while joint is not None:
+            link = joint.parent
+            self._ancestors.insert(0, link)
+            joint = link.parent
+
+    def add_child(self, child: Any):
+        self._children.append(child)
+
+    def remove_child(self, child: Any):
+        self._children.remove(child)
```

Comment:
Would these mess with the `_ancestors` list? Would need to trigger `update_ancestors` but across the entire tree not just for a single class instance which would be more of a global update call that I guess recursively calls all the individual `update_ancestors`.

Reply:
Is this a typical use case? If this just results in some annoying bookkeeping for what turns out to be an uncommon scenario, I'd rather just "lock" links once the ancestors are set for the first time, and prevent users from adding or removing any children.

Reply:
These functions are only used to construct the kinematic tree from urdf. In fact, `_ancestors` are updated at the end so these functions should not induce any bugs for current implementation. But I'm considering making these functions private so that users have no access to them.

Reply:
Might just be my subjective opinion, but I don't think it's safe to have the validity of a state variable depend on correct usage of other parts of the code. Even having a boolean flag like `_is_ancestor_up_to_date`, setting it to `True` in `update_ancestors()`, setting it to `False` in `add_child()`, `remove_child()`, ..etc, and printing out a warning in `ancestors()` if `_is_ancestor_up_to_date is False` would be a good safeguard against incorrect usage of the module.

Reply:
These `setters` are removed for encapsulation.